bitbots_blackboard/src/bitbots_blackboard/capsules/team_data_capsule.py

```diff
@@ -12,53 +12,61 @@ class TeamDataCapsule:
         self.strategy_sender = None  # type: rospy.Publisher
-        self.team_data = TeamData()
+        # indexed with one to match robot ids
+        self.team_data = [None, TeamData(), TeamData(), TeamData(), TeamData(), TeamData(), TeamData()]
         self.team_strategy = dict()
         self.times_to_ball = dict()
         self.strategy = Strategy()
-        self.last_update_team_data = None
         self.strategy_update = None
         self.action_update = None
         self.role_update = None
+        self.data_timeout = rospy.get_param("team_data_timeout", 2)
+        self.ball_min_confidence = rospy.get_param("ball_min_confidence", 0.5)
 
-    def get_team_goalie_ball_position(self):
+    def outdated(self, data: TeamData):
+        return rospy.Time.now() - data.header.stamp > rospy.Duration(self.data_timeout)
+
+    def get_goalie_ball_position(self):
         """Return the ball relative to the goalie
 
         :return a tuple with the relative ball and the last update time
         """
-        i = 0
-        for state in self.team_data.state:
-            if state == Strategy.ROLE_GOALIE:
-                return (self.team_data.ball_relative[i].x, self.team_data.ball_relative[i].y), self.last_update_team_data
-            i += 1
+        for data in self.team_data[1:]:
+            role = data.strategy.role
+            if role == Strategy.ROLE_GOALIE and not self.outdated(data):
+                return data.ball_relative.x, data.ball_relative.y
         return None
 
     def get_goalie_ball_distance(self):
         """Return the distance between the goalie and the ball
 
         :return a tuple with the ball-goalie-distance and the last update time
         """
-        goalie_ball_position = self.get_team_goalie_ball_position()
+        goalie_ball_position = self.get_goalie_ball_position()
         if goalie_ball_position is not None:
             return math.sqrt(goalie_ball_position[0] ** 2 + goalie_ball_position[1] ** 2)
         else:
             return None
 
-    def team_rank_to_ball(self, count_goalies=True):
+    def team_rank_to_ball(self, own_ball_distance, count_goalies=True):
         """Returns the rank of this robot compared to the team robots concerning ball distance.
         Ignores the goalies distance, as it should not leave the goal, even if it is closer than field players.
         For example, we do not want our goalie to perform a throw in against our empty goal.
 
         :return the rank from 1 (nearest) to the number of robots
         """
-        own_time = self.team_data.time_to_position_at_ball
-        sorted_times = dict(sorted(self.times_to_ball.items(), key=lambda item: item[1]))
+        distances = []
+        for data in self.team_data[1:]:
+            if not self.outdated(data) and (data.strategy.role != Strategy.ROLE_GOALIE or count_goalies) \
+                    and data.ball_relative.confidence > self.ball_min_confidence:
+                distances.append(math.sqrt(
```

Comment:
This uses the distance and not the time to the ball, so players that are e.g. not actively playing are counted. This a smaller problem, since penalized players aren't allowed to communicate, but there may be situations that are similar.

This also only includes translation at the moment and not rotation.

Idea: Check if the ball position of the teammate matches our ball position and remove it from the list otherwise. So we still go to our estimation even if a robot has seen a false positive ball.

Reply:
I changed it so that penalized robots are ignored. 

The distance is used instead of time to ball since:
- currently there is no difference as we only scale this by a factor
- this factor is currently defined in team comm config. I don't have a good idea how to make sure that both parts (team_comm and behavior) use the same factor. If they are different it will lead to strange errors. It would be possible to use this param also from behavior, but this would then require to always have this package. 
- maybe the behavior should compute this time and publish it for the team_comm. Then it would also make sense to include a more complex computation using orientation and obstacles in between

Reply:
`data.state != TeamData.STATE_PENALIZED`
As @Flova mentioned above, penalized players don't communicate, so this check should always be true (unless something ignores the rules)

Reply:
> `data.state != TeamData.STATE_PENALIZED`
> As @Flova mentioned above, penalized players don't communicate, so this check should always be true (unless something ignores the rules)

Yes, I would still rather do this sanity check here just to be sure. If at sometime something changes in the rules or if there is a bug in teamcom, then this will not break